File: 42.cpp
#include <vector>
using namespace std;

class Solution {
public:
  int trap(vector<int> &height) {
    int l = 1, r = height.size() - 2;
    int mr = height[height.size() - 1], ml = height[0];
    int ans = 0;
    while (l <= r) {
      if (ml < mr) {
        if (ml < height[l]) {
          ml = height[l];
        } else {
          ans += ml - height[l];
        }
        ++l;
      } else {
        if (mr < height[r]) {
          mr = height[r];
        } else {
          ans += mr - height[r];
        }
        --r;
      }
    }
    return ans;
  }
};

File: 42.py
class Solution:
    def trap(self, height: list[int]) -> int:
        maxL, maxR = height[0], height[len(height) - 1]
        l, r = 0, len(height) - 2
        ans = 0
        while l <= r:
            if maxL < maxR:
                if height[l] < maxL:
                    ans += maxL - height[l]
                else:
                    maxL = height[l]
                l += 1
            else:
                if height[r] < maxR:
                    ans += maxR - height[r]
                else:
                    maxR = height[r]
                r -= 1
        return ans
